File: src-tauri/src/parser/entities/player_data.rs
// Player-nested gameplay data parsers
//
// This module parses gameplay data that is nested within <Player> elements:
// - YieldStockpile → player_resources
// - TechProgress → technology_progress
// - TechCount → technologies_completed
// - TechAvailable/Passed/Trashed/Locked/Target → technology_states
// - CouncilCharacter → player_council
// - ActiveLaw → laws
// - GoalList → player_goals

use crate::parser::id_mapper::IdMapper;
use crate::parser::{ParseError, Result};
use duckdb::{params, Connection};
use roxmltree::Node;

/// Parse YieldStockpile to player_resources table
///
/// Example XML:
/// ```xml
/// <YieldStockpile>
///   <YIELD_CIVICS>15595</YIELD_CIVICS>
///   <YIELD_TRAINING>4874</YIELD_TRAINING>
/// </YieldStockpile>
/// ```
pub fn parse_player_resources(
    player_node: &Node,
    conn: &Connection,
    player_id: i64,
    match_id: i64,
) -> Result<usize> {
    let mut count = 0;

    // Find YieldStockpile child element
    if let Some(yields_node) = player_node.children().find(|n| n.has_tag_name("YieldStockpile")) {
        // Iterate over all child elements (YIELD_CIVICS, YIELD_TRAINING, etc.)
        for yield_node in yields_node.children().filter(|n| n.is_element()) {
            let yield_type = yield_node.tag_name().name(); // "YIELD_CIVICS"
            let amount: i32 = yield_node
                .text()
                .ok_or_else(|| {
                    ParseError::MissingElement(format!(
                        "YieldStockpile.{} text content",
                        yield_type
                    ))
                })?
                .parse()
                .map_err(|_| {
                    ParseError::InvalidFormat(format!(
                        "YieldStockpile.{} must be an integer",
                        yield_type
                    ))
                })?;

            conn.execute(
                "INSERT INTO player_resources (player_id, match_id, yield_type, amount)
                 VALUES (?, ?, ?, ?)",
                params![player_id, match_id, yield_type, amount],
            )?;

            count += 1;
        }
    }

    Ok(count)
}

/// Parse TechProgress to technology_progress table
///
/// Example XML:
/// ```xml
/// <TechProgress>
///   <TECH_IRONWORKING>1001</TECH_IRONWORKING>
///   <TECH_STONECUTTING>414</TECH_STONECUTTING>
/// </TechProgress>
/// ```
pub fn parse_technology_progress(
    player_node: &Node,
    conn: &Connection,
    player_id: i64,
    match_id: i64,
) -> Result<usize> {
    let mut count = 0;

    if let Some(tech_progress_node) = player_node
        .children()
        .find(|n| n.has_tag_name("TechProgress"))
    {
        for tech_node in tech_progress_node.children().filter(|n| n.is_element()) {
            let tech = tech_node.tag_name().name();
            let progress: i32 = tech_node
                .text()
                .ok_or_else(|| {
                    ParseError::MissingElement(format!("TechProgress.{} text content", tech))
                })?
                .parse()
                .map_err(|_| {
                    ParseError::InvalidFormat(format!("TechProgress.{} must be an integer", tech))
                })?;

            conn.execute(
                "INSERT INTO technology_progress (player_id, match_id, tech, progress)
                 VALUES (?, ?, ?, ?)",
                params![player_id, match_id, tech, progress],
            )?;

            count += 1;
        }
    }

    Ok(count)
}

/// Parse TechCount to technologies_completed table
///
/// TechCount contains entries where count > 0 indicates the tech is completed.
///
/// Example XML:
/// ```xml
/// <TechCount>
///   <TECH_IRONWORKING>1</TECH_IRONWORKING>
///   <TECH_TRAPPING>1</TECH_TRAPPING>
/// </TechCount>
/// ```
pub fn parse_technologies_completed(
    player_node: &Node,
    conn: &Connection,
    player_id: i64,
    match_id: i64,
) -> Result<usize> {
    let mut count = 0;

    if let Some(tech_count_node) = player_node.children().find(|n| n.has_tag_name("TechCount")) {
        for tech_node in tech_count_node.children().filter(|n| n.is_element()) {
            let tech = tech_node.tag_name().name();
            let tech_count: i32 = tech_node
                .text()
                .ok_or_else(|| {
                    ParseError::MissingElement(format!("TechCount.{} text content", tech))
                })?
                .parse()
                .map_err(|_| {
                    ParseError::InvalidFormat(format!("TechCount.{} must be an integer", tech))
                })?;

            // Only insert if count > 0 (tech is completed)
            if tech_count > 0 {
                // We don't have completed_turn in the XML at this location
                // It would need to come from a different source or be NULL
                conn.execute(
                    "INSERT INTO technologies_completed (player_id, match_id, tech, completed_turn)
                     VALUES (?, ?, ?, ?)",
                    params![player_id, match_id, tech, None::<i32>],
                )?;

                count += 1;
            }
        }
    }

    Ok(count)
}

/// Parse technology states (TechAvailable, TechPassed, TechTrashed, TechLocked, TechTarget)
///
/// Example XML:
/// ```xml
/// <TechAvailable>
///   <TECH_FORESTRY />
///   <TECH_SAILING />
/// </TechAvailable>
/// <TechPassed>
///   <TECH_MINING />
/// </TechPassed>
/// ```
pub fn parse_technology_states(
    player_node: &Node,
    conn: &Connection,
    player_id: i64,
    match_id: i64,
) -> Result<usize> {
    let mut count = 0;

    // Map of XML element names to state values
    let state_mappings = [
        ("TechAvailable", "available"),
        ("TechPassed", "passed"),
        ("TechTrashed", "trashed"),
        ("TechLocked", "locked"),
        ("TechTarget", "targeted"),
    ];

    for (element_name, state) in &state_mappings {
        if let Some(state_node) = player_node.children().find(|n| n.has_tag_name(*element_name)) {
            for tech_node in state_node.children().filter(|n| n.is_element()) {
                let tech = tech_node.tag_name().name();

                conn.execute(
                    "INSERT INTO technology_states (player_id, match_id, tech, state)
                     VALUES (?, ?, ?, ?)",
                    params![player_id, match_id, tech, state],
                )?;

                count += 1;
            }
        }
    }

    Ok(count)
}

/// Parse CouncilCharacter to player_council table
///
/// Example XML:
/// ```xml
/// <CouncilCharacter>
///   <COUNCIL_GRAND_VIZIER>12</COUNCIL_GRAND_VIZIER>
///   <COUNCIL_CHANCELLOR>5</COUNCIL_CHANCELLOR>
/// </CouncilCharacter>
/// ```
pub fn parse_player_council(
    player_node: &Node,
    conn: &Connection,
    id_mapper: &IdMapper,
    player_id: i64,
    match_id: i64,
) -> Result<usize> {
    let mut count = 0;

    if let Some(council_node) = player_node
        .children()
        .find(|n| n.has_tag_name("CouncilCharacter"))
    {
        for position_node in council_node.children().filter(|n| n.is_element()) {
            let position = position_node.tag_name().name(); // COUNCIL_GRAND_VIZIER
            let character_xml_id: i32 = position_node
                .text()
                .ok_or_else(|| {
                    ParseError::MissingElement(format!(
                        "CouncilCharacter.{} text content",
                        position
                    ))
                })?
                .parse()
                .map_err(|_| {
                    ParseError::InvalidFormat(format!(
                        "CouncilCharacter.{} must be an integer",
                        position
                    ))
                })?;

            // Map character XML ID to database ID
            let character_id = id_mapper.get_character(character_xml_id)?;

            // We don't have appointed_turn in this XML location
            conn.execute(
                "INSERT INTO player_council (player_id, match_id, position, character_id, appointed_turn)
                 VALUES (?, ?, ?, ?, ?)",
                params![player_id, match_id, position, character_id, None::<i32>],
            )?;

            count += 1;
        }
    }

    Ok(count)
}

/// Parse ActiveLaw to laws table
///
/// Example XML:
/// ```xml
/// <ActiveLaw>
///   <LAWCLASS_ORDER>LAW_PRIMOGENITURE</LAWCLASS_ORDER>
///   <LAWCLASS_EPICS_EXPLORATION>LAW_EXPLORATION</LAWCLASS_EPICS_EXPLORATION>
/// </ActiveLaw>
/// ```
pub fn parse_laws(
    player_node: &Node,
    conn: &Connection,
    player_id: i64,
    match_id: i64,
) -> Result<usize> {
    let mut count = 0;

    if let Some(active_law_node) = player_node.children().find(|n| n.has_tag_name("ActiveLaw")) {
        for law_node in active_law_node.children().filter(|n| n.is_element()) {
            let law_category = law_node.tag_name().name(); // LAWCLASS_ORDER
            let law = law_node
                .text()
                .ok_or_else(|| {
                    ParseError::MissingElement(format!("ActiveLaw.{} text content", law_category))
                })?;

            // We don't have adopted_turn in ActiveLaw
            // Use 0 as placeholder (schema requires NOT NULL)
            // Change count might be in LawClassChangeCount
            conn.execute(
                "INSERT INTO laws (player_id, match_id, law_category, law, adopted_turn, change_count)
                 VALUES (?, ?, ?, ?, ?, ?)",
                params![player_id, match_id, law_category, law, 0, 1],
            )?;

            count += 1;
        }
    }

    Ok(count)
}

/// Parse GoalList to player_goals table
///
/// Example XML:
/// ```xml
/// <GoalList>
///   <GoalData>
///     <Type>GOAL_SIX_TECHS</Type>
///     <ID>0</ID>
///     <LeaderID>4</LeaderID>
///     <Turn>37</Turn>
///     <MaxTurns>20</MaxTurns>
///     <Legacy />
///     <Stats>
///       <STAT_TECH_DISCOVERED>5</STAT_TECH_DISCOVERED>
///     </Stats>
///   </GoalData>
/// </GoalList>
/// ```
pub fn parse_player_goals(
    player_node: &Node,
    conn: &Connection,
    id_mapper: &IdMapper,
    player_id: i64,
    match_id: i64,
) -> Result<usize> {
    let mut count = 0;

    if let Some(goal_list_node) = player_node.children().find(|n| n.has_tag_name("GoalList")) {
        for goal_data_node in goal_list_node.children().filter(|n| n.has_tag_name("GoalData")) {
            // Required fields
            let goal_type: String = goal_data_node
                .children()
                .find(|n| n.has_tag_name("Type"))
                .and_then(|n| n.text())
                .ok_or_else(|| ParseError::MissingElement("GoalData.Type".to_string()))?
                .to_string();

            let goal_xml_id: i32 = goal_data_node
                .children()
                .find(|n| n.has_tag_name("ID"))
                .and_then(|n| n.text())
                .ok_or_else(|| ParseError::MissingElement("GoalData.ID".to_string()))?
                .parse()
                .map_err(|_| ParseError::InvalidFormat("GoalData.ID must be an integer".to_string()))?;

            let started_turn: i32 = goal_data_node
                .children()
                .find(|n| n.has_tag_name("Turn"))
                .and_then(|n| n.text())
                .ok_or_else(|| ParseError::MissingElement("GoalData.Turn".to_string()))?
                .parse()
                .map_err(|_| ParseError::InvalidFormat("GoalData.Turn must be an integer".to_string()))?;

            // Optional fields
            let leader_xml_id: Option<i32> = goal_data_node
                .children()
                .find(|n| n.has_tag_name("LeaderID"))
                .and_then(|n| n.text())
                .and_then(|s| s.parse().ok());

            let leader_character_id: Option<i64> = if let Some(xml_id) = leader_xml_id {
                Some(id_mapper.get_character(xml_id)?)
            } else {
                None
            };

            let max_turns: Option<i32> = goal_data_node
                .children()
                .find(|n| n.has_tag_name("MaxTurns"))
                .and_then(|n| n.text())
                .and_then(|s| s.parse().ok());

            // Check if goal is finished or failed
            let finished = goal_data_node
                .children()
                .any(|n| n.has_tag_name("Finished"));

            let completed_turn: Option<i32> = if finished {
                Some(started_turn) // We don't have exact completion turn, use started_turn as placeholder
            } else {
                None
            };

            let failed_turn: Option<i32> = None; // Would need to check GoalsFailed or similar

            // Generate a stable goal_id based on player, match, and goal XML ID
            let goal_id = (match_id as i64) * 1_000_000 + (player_id as i64) * 1000 + (goal_xml_id as i64);

            // Serialize Stats to JSON if present
            let goal_state: Option<String> = if let Some(stats_node) = goal_data_node
                .children()
                .find(|n| n.has_tag_name("Stats"))
            {
                let mut stats = serde_json::Map::new();
                for stat_node in stats_node.children().filter(|n| n.is_element()) {
                    let stat_name = stat_node.tag_name().name();
                    if let Some(stat_value) = stat_node.text() {
                        stats.insert(
                            stat_name.to_string(),
                            serde_json::Value::String(stat_value.to_string()),
                        );
                    }
                }
                if !stats.is_empty() {
                    serde_json::to_string(&stats).ok()
                } else {
                    None
                }
            } else {
                None
            };

            conn.execute(
                "INSERT INTO player_goals (
                    goal_id, player_id, match_id, goal_type, leader_character_id,
                    started_turn, completed_turn, failed_turn, max_turns, progress, goal_state
                )
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                params![
                    goal_id,
                    player_id,
                    match_id,
                    goal_type,
                    leader_character_id,
                    started_turn,
                    completed_turn,
                    failed_turn,
                    max_turns,
                    0, // progress - would need to calculate from Stats
                    goal_state
                ],
            )?;

            count += 1;
        }
    }

    Ok(count)
}

#[cfg(test)]
mod tests {
    use crate::parser::xml_loader::parse_xml;

    #[test]
    fn test_parse_yield_stockpile() {
        let xml = r#"<Root GameId="test">
            <Player ID="0" Name="Test">
                <YieldStockpile>
                    <YIELD_CIVICS>15595</YIELD_CIVICS>
                    <YIELD_TRAINING>4874</YIELD_TRAINING>
                </YieldStockpile>
            </Player>
        </Root>"#;

        let doc = parse_xml(xml.to_string()).unwrap();
        let player_node = doc
            .root_element()
            .children()
            .find(|n| n.has_tag_name("Player"))
            .unwrap();

        let yields_node = player_node
            .children()
            .find(|n| n.has_tag_name("YieldStockpile"))
            .unwrap();

        let mut count = 0;
        for yield_node in yields_node.children().filter(|n| n.is_element()) {
            assert!(yield_node.tag_name().name().starts_with("YIELD_"));
            assert!(yield_node.text().is_some());
            count += 1;
        }

        assert_eq!(count, 2);
    }
}

File: src/lib/parser/parsers/player-data.ts
// Player-nested data parsers. Seven collections per <Player>: resources,
// technology progress/completed/states, council, laws, goals.

import { ParseError } from "../extract-zip.js";
import {
	asArray,
	collectStrictNamedInts,
	getElementChildren,
	isElement,
	optInt,
	optStr,
	requireInt,
} from "../parse-xml.js";

// ---------- Types ----------

export interface PlayerResource {
	playerXmlId: number;
	yieldType: string;
	amount: number;
}

export interface TechnologyProgress {
	playerXmlId: number;
	tech: string;
	progress: number;
}

export interface TechnologyCompleted {
	playerXmlId: number;
	tech: string;
	completedTurn: number | null;
}

// One tech a player chose, with the others that were on offer in the same
// draw. See parseTechChoices.
export interface TechChoice {
	playerXmlId: number;
	tech: string;
	alternates: string[];
}

export interface TechnologyState {
	playerXmlId: number;
	tech: string;
	state: string;
}

export interface PlayerCouncil {
	playerXmlId: number;
	position: string;
	characterXmlId: number;
	appointedTurn: number | null;
}

export interface Law {
	playerXmlId: number;
	lawCategory: string;
	law: string;
	adoptedTurn: number;
	changeCount: number;
}

export interface PlayerGoal {
	playerXmlId: number;
	goalXmlId: number;
	goalType: string;
	leaderCharacterXmlId: number | null;
	startedTurn: number;
	completedTurn: number | null;
	failedTurn: number | null;
	maxTurns: number | null;
	progress: number;
	goalState: string | null;
}

// ---------- Helpers ----------

function* eachPlayer(
	root: Record<string, unknown>,
): Generator<[number, Record<string, unknown>]> {
	for (const node of asArray(root.Player) as unknown[]) {
		if (!isElement(node)) continue;
		const playerXmlId = requireInt(node["@_ID"], "Player.ID");
		yield [playerXmlId, node];
	}
}

// State name mapping for parseTechnologyStates. Order matters: Rust walks
// these in this exact sequence.
const TECH_STATE_MAPPINGS: ReadonlyArray<[string, string]> = [
	["TechAvailable", "available"],
	["TechPassed", "passed"],
	["TechTrashed", "trashed"],
	["TechLocked", "locked"],
	// Note: TechTarget → "targeted" (not "target") per Rust source.
	["TechTarget", "targeted"],
];

// ---------- Player resources ----------

export function parsePlayerResources(
	root: Record<string, unknown>,
): PlayerResource[] {
	const out: PlayerResource[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		for (const { name, value } of collectStrictNamedInts(
			node.YieldStockpile,
			"YieldStockpile",
		)) {
			out.push({ playerXmlId, yieldType: name, amount: value });
		}
	}
	return out;
}

// ---------- Projects produced (Player.ProjectsProduced) ----------
//
// Name-keyed int map on the Player node — every project the player ever
// completed, with counts, city projects and player projects alike. The
// per-City <ProjectCount> carries the same data split by city; this is the
// authoritative whole-game total (a city lost to conquest keeps its projects
// here but not there). No Rust counterpart: first parsed at 2.13.0.

export interface PlayerProjectProduced {
	playerXmlId: number;
	project: string;
	count: number;
}

export function parseProjectsProduced(
	root: Record<string, unknown>,
): PlayerProjectProduced[] {
	const out: PlayerProjectProduced[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		for (const { name, value } of collectStrictNamedInts(
			node.ProjectsProduced,
			"ProjectsProduced",
		)) {
			out.push({ playerXmlId, project: name, count: value });
		}
	}
	return out;
}

// ---------- Technology progress ----------

export function parseTechnologyProgress(
	root: Record<string, unknown>,
): TechnologyProgress[] {
	const out: TechnologyProgress[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		for (const { name, value } of collectStrictNamedInts(
			node.TechProgress,
			"TechProgress",
		)) {
			out.push({ playerXmlId, tech: name, progress: value });
		}
	}
	return out;
}

// ---------- Technologies completed ----------

export function parseTechnologiesCompleted(
	root: Record<string, unknown>,
): TechnologyCompleted[] {
	const out: TechnologyCompleted[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		for (const { name, value } of collectStrictNamedInts(
			node.TechCount,
			"TechCount",
		)) {
			// Rust filter: only emit if count > 0. Skip TECH_X with 0.
			if (value <= 0) continue;
			out.push({
				playerXmlId,
				tech: name,
				// Hardcoded null — not available at this XML location.
				completedTurn: null,
			});
		}
	}
	return out;
}

// ---------- Technology states ----------

export function parseTechnologyStates(
	root: Record<string, unknown>,
): TechnologyState[] {
	const out: TechnologyState[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		for (const [elementName, state] of TECH_STATE_MAPPINGS) {
			const stateNode = node[elementName];
			if (!isElement(stateNode)) continue;
			// Just collect element child names — values are ignored (children
			// are typically self-closing like <TECH_FORESTRY/>).
			for (const [tech] of getElementChildren(stateNode)) {
				out.push({ playerXmlId, tech, state });
			}
		}
	}
	return out;
}

// ---------- Tech choices (Player.TechPathHistory) ----------
//
// Old World deals a hand of techs and the player takes one; the rest stay in
// the deck. <TechPathHistory> is the record of those draws, one child element
// per (chosen, passed-over) pair, named for the pair itself:
//
//   <TechPathHistory>
//     <TECH_IRONWORKING.TECH_DRAMA />
//     <TECH_IRONWORKING.TECH_MILITARY_DRILL />
//
// reading as "when Ironworking was taken, Drama and Military Drill were the
// other cards". Element ORDER is the order the draws happened in, and is
// preserved — it is the only ordering the save gives, since no turn is
// recorded against a choice.
//
// A tech the player holds that appears as no key here was never drafted: it
// was granted (a nation's starting techs, a Sages seat's random tech, an event
// reward). The consumer separates those using the baked starting-tech table.
//
// The dot in the element name is what makes this parseable at all — tech
// zTypes never contain one, so a single split recovers both halves. A name
// that doesn't split in two is skipped rather than guessed at.

export function parseTechChoices(root: Record<string, unknown>): TechChoice[] {
	const out: TechChoice[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		const history = node.TechPathHistory;
		if (!isElement(history)) continue;
		// Keyed by chosen tech, insertion-ordered — the draws come out in the
		// order they happened.
		const byTech = new Map<string, string[]>();
		for (const [name] of getElementChildren(history)) {
			const dot = name.indexOf(".");
			if (dot <= 0 || dot === name.length - 1) continue;
			const chosen = name.slice(0, dot);
			const alternate = name.slice(dot + 1);
			if (alternate.includes(".")) continue;
			const alternates = byTech.get(chosen) ?? [];
			alternates.push(alternate);
			byTech.set(chosen, alternates);
		}
		for (const [tech, alternates] of byTech) {
			out.push({ playerXmlId, tech, alternates });
		}
	}
	return out;
}

// ---------- Player council ----------

export function parsePlayerCouncil(
	root: Record<string, unknown>,
): PlayerCouncil[] {
	const out: PlayerCouncil[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		for (const { name, value } of collectStrictNamedInts(
			node.CouncilCharacter,
			"CouncilCharacter",
		)) {
			out.push({
				playerXmlId,
				position: name,
				characterXmlId: value,
				// Hardcoded null — not available at this XML location.
				appointedTurn: null,
			});
		}
	}
	return out;
}

// ---------- Laws ----------

export function parseLaws(root: Record<string, unknown>): Law[] {
	const out: Law[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		const activeLawNode = node.ActiveLaw;
		if (!isElement(activeLawNode)) continue;

		// Strict named-string container. Each child name → law_category,
		// text → law (e.g. <LAWCLASS_ORDER>LAW_PRIMOGENITURE</LAWCLASS_ORDER>).
		// Inline rather than promoting another helper — single use.
		for (const [lawCategory, value] of getElementChildren(activeLawNode)) {
			if (typeof value !== "string" || value === "") {
				throw new ParseError(
					`ActiveLaw.${lawCategory} text content`,
					"MISSING_FIELD",
				);
			}
			out.push({
				playerXmlId,
				lawCategory,
				law: value,
				// Literal int placeholders — the parse layer doesn't have
				// adoption-turn or change-count data. NOT null. Match Rust.
				adoptedTurn: 0,
				changeCount: 1,
			});
		}
	}
	return out;
}

// ---------- Player goals ----------

export function parsePlayerGoals(root: Record<string, unknown>): PlayerGoal[] {
	const out: PlayerGoal[] = [];
	for (const [playerXmlId, node] of eachPlayer(root)) {
		const goalListNode = node.GoalList;
		if (!isElement(goalListNode)) continue;

		for (const goalData of asArray(goalListNode.GoalData) as unknown[]) {
			if (!isElement(goalData)) continue;

			// Required fields — error if missing.
			const goalType = optStr(goalData.Type);
			if (goalType === null) {
				throw new ParseError("GoalData.Type", "MISSING_FIELD");
			}

			const goalIdRaw = optStr(goalData.ID);
			if (goalIdRaw === null) {
				throw new ParseError("GoalData.ID", "MISSING_FIELD");
			}
			const goalXmlId = parseInt(goalIdRaw, 10);
			if (Number.isNaN(goalXmlId)) {
				throw new ParseError(
					"GoalData.ID must be an integer",
					"INVALID_FORMAT",
				);
			}

			const turnRaw = optStr(goalData.Turn);
			if (turnRaw === null) {
				throw new ParseError("GoalData.Turn", "MISSING_FIELD");
			}
			const startedTurn = parseInt(turnRaw, 10);
			if (Number.isNaN(startedTurn)) {
				throw new ParseError(
					"GoalData.Turn must be an integer",
					"INVALID_FORMAT",
				);
			}

			// Optional fields — lenient parse.
			const leaderCharacterXmlId = optInt(goalData.LeaderID);
			const maxTurns = optInt(goalData.MaxTurns);

			// <Finished/> presence check. If present, completed_turn =
			// started_turn (literal "use started_turn as placeholder" hack
			// hack). NOT a separate read.
			const finished = "Finished" in goalData;
			const completedTurn = finished ? startedTurn : null;

			out.push({
				playerXmlId,
				goalXmlId,
				goalType,
				leaderCharacterXmlId,
				startedTurn,
				completedTurn,
				// Hardcoded — not available at this layer.
				failedTurn: null,
				maxTurns,
				progress: 0,
				goalState: serializeGoalStats(goalData.Stats),
			});
		}
	}
	return out;
}

/**
 * Walk `<Stats>` element children into a JSON-encoded string, with keys
 * sorted alphabetically to match Rust's `serde_json::Map` (default features
 * → `BTreeMap` → byte-lex key sort). Returns null when Stats is absent or
 * resolves to an empty map (mirrors Rust `if !stats.is_empty()` check).
 *
 * Empty-text children are silently skipped (mirrors Rust's
 * `if let Some(text)` short-circuit; fast-xml-parser surfaces empty
 * elements as `""`).
 *
 * Output format must match Rust serde_json::to_string exactly: compact
 * (no spaces), all string values, keys ASCII-sorted.
 */
function serializeGoalStats(node: unknown): string | null {
	if (!isElement(node)) return null;
	const stats: Record<string, string> = {};
	for (const [statName, statValue] of getElementChildren(node)) {
		if (typeof statValue !== "string" || statValue === "") continue;
		stats[statName] = statValue;
	}
	const keys = Object.keys(stats);
	if (keys.length === 0) return null;
	// Replacer-array second arg gates the keys AND fixes their order. For
	// our keys (ASCII), Array.prototype.sort() matches Rust's BTreeMap<String>
	// byte-lex ordering.
	return JSON.stringify(stats, keys.sort());
}

// ---------- ToRow mappers (snake_case wire format) ----------

export function playerResourceToRow(
	r: PlayerResource,
): Record<string, unknown> {
	return {
		player_xml_id: r.playerXmlId,
		yield_type: r.yieldType,
		amount: r.amount,
	};
}

export function techChoiceToRow(c: TechChoice): Record<string, unknown> {
	return {
		player_xml_id: c.playerXmlId,
		tech: c.tech,
		alternates: c.alternates,
	};
}

export function projectProducedToRow(
	r: PlayerProjectProduced,
): Record<string, unknown> {
	return {
		player_xml_id: r.playerXmlId,
		project: r.project,
		count: r.count,
	};
}

export function playerGoalToRow(g: PlayerGoal): Record<string, unknown> {
	return {
		player_xml_id: g.playerXmlId,
		goal_xml_id: g.goalXmlId,
		goal_type: g.goalType,
		leader_character_xml_id: g.leaderCharacterXmlId,
		started_turn: g.startedTurn,
		completed_turn: g.completedTurn,
		failed_turn: g.failedTurn,
		max_turns: g.maxTurns,
		progress: g.progress,
		goal_state: g.goalState,
	};
}
